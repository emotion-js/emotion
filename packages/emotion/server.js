module.exports = require('emotion-server')
