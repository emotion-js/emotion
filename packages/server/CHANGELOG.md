# @emotion/server

## 11.4.0

### Minor Changes

- [#2334](https://github.com/emotion-js/emotion/pull/2334) [`7d9e74f8`](https://github.com/emotion-js/emotion/commit/7d9e74f8f0d0e0ea60573a19965eded61dc41024) Thanks [@mnajdova](https://github.com/mnajdova)! - Added `extractCriticalToChunks` that allows the Global component to remove styles on unmount unlike `extractCritical` along with `constructStyleTagsFromChunks` to render the chunks to style tags.

## 11.0.0

### Major Changes

- [`b8476e08`](https://github.com/emotion-js/emotion/commit/b8476e08af4a2e8de94a112cb0daf6e8e4d56fe1) [#1675](https://github.com/emotion-js/emotion/pull/1675) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Rename `emotion-server` to `@emotion/server`. Please change any imports of `emotion-server` to import `@emotion/server` or use the `@emotion/pkg-renaming` ESLint rule from `@emotion/eslint-plugin`.

* [`b8476e08`](https://github.com/emotion-js/emotion/commit/b8476e08af4a2e8de94a112cb0daf6e8e4d56fe1) [#1675](https://github.com/emotion-js/emotion/pull/1675) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Move `create-emotion-server` to `@emotion/server/create-instance`. Please change any imports of `create-emotion-server` to import `@emotion/server/create-instance` or use the `@emotion/pkg-renaming` ESLint rule from `@emotion/eslint-plugin`.

* [`105de5c8`](https://github.com/emotion-js/emotion/commit/105de5c8752be0983c000e1e26462dc8fcf0708d) [#1572](https://github.com/emotion-js/emotion/pull/1572) Thanks [@Andarist](https://github.com/Andarist)! - `[data-emotion]` attribute on SSRed styled has changed. You should never rely on it though.

### Patch Changes

- Updated dependencies [[`debaad9a`](https://github.com/emotion-js/emotion/commit/debaad9ab4bd6c80312092826d9146f3d29c0899), [`9e998e37`](https://github.com/emotion-js/emotion/commit/9e998e3755c217027ad1be0af4c64644fe14c6bf)]:
  - @emotion/utils@1.0.0

## 11.0.0-rc.0

### Major Changes

- [`9c4ebc16`](https://github.com/emotion-js/emotion/commit/9c4ebc160471097c5d04fb92dba3ed0df870bb63) [#2030](https://github.com/emotion-js/emotion/pull/2030) Thanks [@Andarist](https://github.com/Andarist)! - Release candidate version.

### Patch Changes

- Updated dependencies [[`9c4ebc16`](https://github.com/emotion-js/emotion/commit/9c4ebc160471097c5d04fb92dba3ed0df870bb63)]:
  - @emotion/css@11.0.0-rc.0
  - @emotion/utils@1.0.0-rc.0

## 11.0.0-next.19

### Patch Changes

- Updated dependencies []:
  - @emotion/css@11.0.0-next.19

## 11.0.0-next.17

### Patch Changes

- Updated dependencies []:
  - @emotion/css@11.0.0-next.17

## 11.0.0-next.16

### Patch Changes

- Updated dependencies [[`debaad9a`](https://github.com/emotion-js/emotion/commit/debaad9ab4bd6c80312092826d9146f3d29c0899)]:
  - @emotion/utils@1.0.0-next.1
  - @emotion/css@11.0.0-next.16

## 11.0.0-next.15

### Patch Changes

- Updated dependencies [[`5d692a6a`](https://github.com/emotion-js/emotion/commit/5d692a6a8102b3faabefb773dd0145b123668a07)]:
  - @emotion/css@11.0.0-next.15

## 11.0.0-next.14

### Patch Changes

- Updated dependencies [[`6d32d82b`](https://github.com/emotion-js/emotion/commit/6d32d82beb45b18e5f18a37932b862ad19b17044)]:
  - @emotion/css@11.0.0-next.14

## 11.0.0-next.13

### Patch Changes

- Updated dependencies [[`5e803106`](https://github.com/emotion-js/emotion/commit/5e803106d391b7c036bdf634318b80337a1d9b70), [`9e998e37`](https://github.com/emotion-js/emotion/commit/9e998e3755c217027ad1be0af4c64644fe14c6bf), [`9e998e37`](https://github.com/emotion-js/emotion/commit/9e998e3755c217027ad1be0af4c64644fe14c6bf), [`9e998e37`](https://github.com/emotion-js/emotion/commit/9e998e3755c217027ad1be0af4c64644fe14c6bf)]:
  - @emotion/babel-plugin@11.0.0-next.13
  - @emotion/css@11.0.0-next.13
  - @emotion/utils@1.0.0-next.0

## 11.0.0-next.12

### Major Changes

- [`105de5c8`](https://github.com/emotion-js/emotion/commit/105de5c8752be0983c000e1e26462dc8fcf0708d) [#1572](https://github.com/emotion-js/emotion/pull/1572) Thanks [@Andarist](https://github.com/Andarist)! - `[data-emotion]` attribute on SSRed styled has changed. You should never rely on it though.

### Patch Changes

- Updated dependencies [[`5c7ec859`](https://github.com/emotion-js/emotion/commit/5c7ec85904633a11185066fa591dc8969f3f2ff2), [`5bea60b1`](https://github.com/emotion-js/emotion/commit/5bea60b1ffab85fbc965532006c3a94ea139f0bf), [`e3d7db87`](https://github.com/emotion-js/emotion/commit/e3d7db87deaac95817404760112417ac1fa1b56d), [`105de5c8`](https://github.com/emotion-js/emotion/commit/105de5c8752be0983c000e1e26462dc8fcf0708d)]:
  - @emotion/babel-plugin@11.0.0-next.12
  - @emotion/css@11.0.0-next.12

## 11.0.0-next.10

### Major Changes

- [`b8476e08`](https://github.com/emotion-js/emotion/commit/b8476e08af4a2e8de94a112cb0daf6e8e4d56fe1) [#1675](https://github.com/emotion-js/emotion/pull/1675) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Rename `emotion-server` to `@emotion/server`. Please change any imports of `emotion-server` to import `@emotion/server` or use the `@emotion/pkg-renaming` ESLint rule from `@emotion/eslint-plugin`.

* [`b8476e08`](https://github.com/emotion-js/emotion/commit/b8476e08af4a2e8de94a112cb0daf6e8e4d56fe1) [#1675](https://github.com/emotion-js/emotion/pull/1675) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Move `create-emotion-server` to `@emotion/server/create-instance`. Please change any imports of `create-emotion-server` to import `@emotion/server/create-instance` or use the `@emotion/pkg-renaming` ESLint rule from `@emotion/eslint-plugin`.

### Patch Changes

- Updated dependencies [[`b8476e08`](https://github.com/emotion-js/emotion/commit/b8476e08af4a2e8de94a112cb0daf6e8e4d56fe1), [`b8476e08`](https://github.com/emotion-js/emotion/commit/b8476e08af4a2e8de94a112cb0daf6e8e4d56fe1), [`c7850e61`](https://github.com/emotion-js/emotion/commit/c7850e61211d6aa26a3388399889a6072ee2f1fe), [`b8476e08`](https://github.com/emotion-js/emotion/commit/b8476e08af4a2e8de94a112cb0daf6e8e4d56fe1)]:
  - @emotion/babel-plugin@11.0.0-next.10
  - @emotion/css@11.0.0-next.10
  - emotion@11.0.0-next.10

## 11.0.0-next.0

### Major Changes

- [`302bdba1`](https://github.com/emotion-js/emotion/commit/302bdba1a6b793484c09edeb668815c5e31ea555) [#1600](https://github.com/emotion-js/emotion/pull/1600) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Ensure packages are major bumped so that pre-release versions of the linked packages are consistent in the major number

### Patch Changes

- Updated dependencies [[`b0ad4f0c`](https://github.com/emotion-js/emotion/commit/b0ad4f0c628813a42c4637857be9a969429db6f0), [`302bdba1`](https://github.com/emotion-js/emotion/commit/302bdba1a6b793484c09edeb668815c5e31ea555)]:
  - babel-plugin-emotion@11.0.0-next.0
  - create-emotion-server@11.0.0-next.0
  - emotion@11.0.0-next.0

## 10.0.27

### Patch Changes

- [`4c62ae9`](https://github.com/emotion-js/emotion/commit/4c62ae9447959d438928e1a26f76f1487983c968) [#1698](https://github.com/emotion-js/emotion/pull/1698) Thanks [@Andarist](https://github.com/Andarist)! - Add LICENSE file
- Updated dependencies [[`4c62ae9`](https://github.com/emotion-js/emotion/commit/4c62ae9447959d438928e1a26f76f1487983c968)]:
  - babel-plugin-emotion@10.0.27
  - create-emotion-server@10.0.27
  - emotion@10.0.27

## 10.0.17

### Patch Changes

- [16ff2330](https://github.com/emotion-js/emotion/commit/16ff233061e35fe71bfb1671da54ac12d6fc9eeb) [#1483](https://github.com/emotion-js/emotion/pull/1483) Thanks [@JakeGinnivan](https://github.com/JakeGinnivan)! - nth selector will no longer warn when using extract critical

## 10.0.14

### Patch Changes

- [c0eb604d](https://github.com/emotion-js/emotion/commit/c0eb604d) [#1419](https://github.com/emotion-js/emotion/pull/1419) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Update build tool
