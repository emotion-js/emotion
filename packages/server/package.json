{
  "name": "@emotion/server",
  "version": "11.4.0",
  "description": "Extract and inline critical css with emotion for server side rendering.",
  "main": "dist/emotion-server.cjs.js",
  "types": "types/index.d.ts",
  "files": [
    "src",
    "dist",
    "types/*.d.ts",
    "create-instance"
  ],
  "scripts": {
    "test:typescript": "dtslint types"
  },
  "dependencies": {
    "@emotion/utils": "^1.0.0",
    "html-tokenize": "^2.0.0",
    "multipipe": "^1.0.2",
    "through": "^2.3.8"
  },
  "peerDependencies": {
    "@emotion/css": "^11.0.0-rc.0"
  },
  "peerDependenciesMeta": {
    "@emotion/css": {
      "optional": true
    }
  },
  "devDependencies": {
    "@emotion/babel-plugin": "11.3.0",
    "@emotion/css": "11.1.3",
    "@emotion/css-prettifier": "1.0.0",
    "@types/node": "^10.11.4",
    "dtslint": "^0.3.0"
  },
  "author": "Kye Hohenberger",
  "homepage": "https://emotion.sh",
  "license": "MIT",
  "repository": "https://github.com/emotion-js/emotion/tree/main/packages/server",
  "keywords": [
    "styles",
    "emotion",
    "react",
    "css",
    "css-in-js",
    "ssr",
    "server-side-rendering"
  ],
  "bugs": {
    "url": "https://github.com/emotion-js/emotion/issues"
  },
  "browser": {
    "./dist/emotion-server.cjs.js": "./dist/emotion-server.browser.cjs.js"
  },
  "preconstruct": {
    "entrypoints": [
      "./index.js",
      "./create-instance/index.js"
    ]
  }
}
