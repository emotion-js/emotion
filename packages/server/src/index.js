// @flow
import createEmotionServer from './create-instance'
import { cache } from '@emotion/css'

export const {
  extractCritical,
  extractCriticalToChunks,
  renderStylesToString,
  renderStylesToNodeStream,
  constructStyleTagsFromChunks
} = createEmotionServer(cache)
