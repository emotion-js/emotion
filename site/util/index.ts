export * from './style-constants'
export * from './dracula-prism'
