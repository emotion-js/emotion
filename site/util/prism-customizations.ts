import Prism from 'prismjs'
import 'prismjs/components/prism-jsx'
import 'prismjs/components/prism-typescript'
import 'prismjs/components/prism-tsx'

// Prevent Prism from automatically highlighting code, which is unnecessary and
// causes hydration warnings.
;(Prism as { manual?: boolean }).manual = true

// This is only necessary for the live editors, for some reason it "just works"
// with remark-prism. But, we make sure this is imported on every page so that
// the Prism configuration doesn't depend on whether there is a live editor
// present.
Prism.languages.insertBefore('jsx', 'template-string', {
  'styled-template-string': {
    pattern:
      /(styled(\.\w+|\([^)]*\))(\.\w+(\([^)]*\))*)*|css|injectGlobal|keyframes|css={)`(?:\$\{[^}]+\}|\\\\|\\?[^\\])*?`/,
    lookbehind: true,
    greedy: true,
    inside: {
      interpolation: {
        pattern: /\$\{[^}]+\}/,
        inside: {
          'interpolation-punctuation': {
            pattern: /^\$\{|\}$/,
            alias: 'punctuation'
          },
          rest: Prism.languages.jsx
        }
      },
      string: {
        pattern: /[^$;]+/,
        inside: Prism.languages.css,
        alias: 'language-css'
      }
    }
  }
})
