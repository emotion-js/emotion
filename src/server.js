import { sheet, inserted } from './index'
import { keys, forEach } from './utils'

const RGX = /css(?:[a-zA-Z0-9-]*)-([a-zA-Z0-9]+)/gm

export {
  flush,
  css,
  injectGlobal,
  fontFace,
  keyframes,
  hydrate,
  objStyle
} from './index'

export function extractCritical (html) {
  // parse out ids from html
  // reconstruct css/rules/cache to pass
  let o = { html, ids: [], css: '', rules: [] }
  let match
  let ids = {}
  while ((match = RGX.exec(html)) !== null) {
    if (!ids[match[1]]) {
      ids[match[1]] = true
    }
  }

  o.rules = sheet.sheet.cssRules.slice().filter(x => {
    let match = RGX.exec(x.cssText)
    return match == null || ids[match[1]] || false
  })

  o.ids = keys(inserted).filter(id => !!ids[id])

  let css = ''
  forEach(o.rules, x => (css += x.cssText))
  o.css = css

  return o
}
