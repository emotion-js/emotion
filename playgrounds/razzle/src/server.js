// @flow
import App from './App'
import * as React from 'react'
import { StaticRouter } from 'react-router-dom'
import express from 'express'
import { renderToString } from 'react-dom/server'
import { renderStylesToString } from 'emotion-server'
import { cache } from 'emotion'
import { CacheProvider } from '@emotion/core'

// $FlowFixMe
const assets = require(process.env.RAZZLE_ASSETS_MANIFEST)

const server = express()
server
  .disable('x-powered-by')
  .use(express.static(process.env.RAZZLE_PUBLIC_DIR))
  .get('/*', (req, res) => {
    const context = {}
    const markup = renderStylesToString(
      renderToString(
        <CacheProvider value={cache}>
          <StaticRouter context={context} location={req.url}>
            <App />
          </StaticRouter>
        </CacheProvider>
      )
    )

    if (context.url) {
      res.redirect(context.url)
    } else {
      res.status(200).send(
        `<!doctype html>
    <html lang="">
    <head>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta charSet='utf-8' />
        <title>Welcome to Razzle</title>
        <meta name="viewport" content="width=device-width, initial-scale=1">
        ${
          assets.client.css
            ? `<link rel="stylesheet" href="${assets.client.css}">`
            : ''
        }
        ${
          process.env.NODE_ENV === 'production'
            ? `<script src="${assets.client.js}" defer></script>`
            : `<script src="${assets.client.js}" defer crossorigin></script>`
        }
    </head>
    <body>
        <div id="root">${markup}</div>
    </body>
</html>`
      )
    }
  })

export default server
