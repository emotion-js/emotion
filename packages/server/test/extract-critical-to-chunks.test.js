/**
 * @jest-environment node
 */

import React from 'react'
import { renderToString } from 'react-dom/server'
/* import type { Emotion } from '@emotion/css/create-instance' */
import { prettifyCriticalChunks } from './util'

let emotion = require('@emotion/css')
let reactEmotion = require('@emotion/styled')
let emotionServer = require('@emotion/server')

export const getComponents = (
  emotion /*: Emotion */,
  { default: styled } /*: { default: Function } */
) => {
  let Provider = require('@emotion/react').CacheProvider
  let Global = require('@emotion/react').Global
  let { css } = emotion

  const hoverStyles = css`
    color: hotpink;
    &:hover {
      color: white;
      background-color: lightgray;
      border-color: aqua;
      box-shadow: -15px -15px 0 0 aqua, -30px -30px 0 0 cornflowerblue;
    }
    label: hoverStyles;
  `

  // this is using @emotion/styled/base
  // so the call syntax has to be used
  const Main = styled('main')`
    ${hoverStyles};
    display: flex;
    label: Something_Main;
  `

  const Page1 = () => (
    <Provider value={emotion.cache}>
      <Global styles={{ body: { color: 'white' } }} />
      <Global styles={{ html: { background: 'red' } }} />
      <Main />
    </Provider>
  )

  const Page2 = () => (
    <Provider value={emotion.cache}>
      <Global styles={{ body: { color: 'white' } }} />
      <Main />
    </Provider>
  )
  return { Page1, Page2 }
}

describe('extractCriticalToChunks', () => {
  const { Page1, Page2 } = getComponents(emotion, reactEmotion)

  const page1Critical = emotionServer.extractCriticalToChunks(
    renderToString(<Page1 />)
  )
  const page2Critical = emotionServer.extractCriticalToChunks(
    renderToString(<Page2 />)
  )

  test('returns static css', () => {
    expect(prettifyCriticalChunks(page1Critical)).toMatchSnapshot()
    expect(prettifyCriticalChunks(page2Critical)).toMatchSnapshot()
  })

  test('generates correct style tags using constructStyleTagsFromChunks', () => {
    expect(
      emotionServer.constructStyleTagsFromChunks(page1Critical)
    ).toMatchSnapshot()
    expect(
      emotionServer.constructStyleTagsFromChunks(page2Critical)
    ).toMatchSnapshot()
  })
})
