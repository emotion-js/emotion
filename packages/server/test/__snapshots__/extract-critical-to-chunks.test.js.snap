// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractCriticalToChunks generates correct style tags using constructStyleTagsFromChunks 1`] = `
<style data-emotion="css-global l6h">
  body{color:white;}
</style>
<style data-emotion="css-global 10q49a4">
  html{background:red;}
</style>
<style data-emotion="css fgjrlg-hoverStyles-Something_Main">
  .css-fgjrlg-hoverStyles-Something_Main{color:hotpink;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.css-fgjrlg-hoverStyles-Something_Main:hover{color:white;background-color:lightgray;border-color:aqua;box-shadow:-15px -15px 0 0 aqua,-30px -30px 0 0 cornflowerblue;}
</style>
`;

exports[`extractCriticalToChunks generates correct style tags using constructStyleTagsFromChunks 2`] = `
<style data-emotion="css-global l6h">
  body{color:white;}
</style>
<style data-emotion="css-global 10q49a4">
  html{background:red;}
</style>
<style data-emotion="css fgjrlg-hoverStyles-Something_Main">
  .css-fgjrlg-hoverStyles-Something_Main{color:hotpink;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}.css-fgjrlg-hoverStyles-Something_Main:hover{color:white;background-color:lightgray;border-color:aqua;box-shadow:-15px -15px 0 0 aqua,-30px -30px 0 0 cornflowerblue;}
</style>
`;

exports[`extractCriticalToChunks returns static css 1`] = `
{
  "html": <main class="css-fgjrlg-hoverStyles-Something_Main">
</main>,
  "styles": [
    {
      "css": "body {
  color: white;
}",
      "ids": [
        "l6h",
      ],
      "key": "css-global",
    },
    {
      "css": "html {
  background: red;
}",
      "ids": [
        "10q49a4",
      ],
      "key": "css-global",
    },
    {
      "css": ".css-fgjrlg-hoverStyles-Something_Main {
  color: hotpink;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.css-fgjrlg-hoverStyles-Something_Main:hover {
  color: white;
  background-color: lightgray;
  border-color: aqua;
  box-shadow: -15px -15px 0 0 aqua,-30px -30px 0 0 cornflowerblue;
}",
      "ids": [
        "fgjrlg-hoverStyles-Something_Main",
      ],
      "key": "css",
    },
  ],
}
`;

exports[`extractCriticalToChunks returns static css 2`] = `
{
  "html": <main class="css-fgjrlg-hoverStyles-Something_Main">
</main>,
  "styles": [
    {
      "css": "body {
  color: white;
}",
      "ids": [
        "l6h",
      ],
      "key": "css-global",
    },
    {
      "css": "html {
  background: red;
}",
      "ids": [
        "10q49a4",
      ],
      "key": "css-global",
    },
    {
      "css": ".css-fgjrlg-hoverStyles-Something_Main {
  color: hotpink;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.css-fgjrlg-hoverStyles-Something_Main:hover {
  color: white;
  background-color: lightgray;
  border-color: aqua;
  box-shadow: -15px -15px 0 0 aqua,-30px -30px 0 0 cornflowerblue;
}",
      "ids": [
        "fgjrlg-hoverStyles-Something_Main",
      ],
      "key": "css",
    },
  ],
}
`;
